### PR TITLE
27598 edit page v2 move communication events to proper library

### LINKED ISSUE #27598: Edit Page V2: Move Communication Events To Proper Library
### Parent
#27546 

### Task description
As a developer, I need to reuse the communication code in the headless app (from the libraries) in the VTL rendered pages.

We need to move the communication code outside the React and Nextjs libraries.

Currently, the communication is tightly coupled with these libraries, which makes it impossible to integrate with the VTL rendered pages. 

By moving the communication to a separate JavaScript library, what we are looking to that we can easily call it from the VTL pages using `<script src="" />`

```[tasklist]
### Acceptance Criteria
- [ ] The existing communication code is refactored to be independent of the react and nextjs libraries.
- [ ] The new JavaScript library is created and implemented in the system.
- [ ] The new library allows us to seamlessly integrate the communication functionality with the VTL pages.
- [ ] All existing functionality related to communication remains intact after the update.
- [ ] The new library documentation needs to be on point
```

### External Links
We have an enum `CUSTOMER_ACTIONS` that has all the events that we send from the client to the editor

https://github.com/dotCMS/core/blob/a4e3a77083a57ce2289138dfaf427c585f9adbcc/core-web/libs/sdk/client/src/lib/postMessageToEditor.ts#L7-L33

We have to send a payload, which could be obtained from the client HTML, along with certain tasks. We have utilities available to fetch or find this payload, and we need to make these utilities accessible from the new libraries.

For example: 

https://github.com/dotCMS/core/blob/master/core-web/libs/sdk/react/src/lib/hooks/usePageEditor.tsx#L124-L133

Where we get all the positions for the rows, columns, containers and contentlets in the page. Likewise, this one, we might have some others, we need to move them as well.

And we have another one `NOTIFY_CUSTOMER` that is the one we use to do communication from the editor to the client:

https://github.com/dotCMS/core/blob/a4e3a77083a57ce2289138dfaf427c585f9adbcc/core-web/libs/portlets/edit-ema/portlet/src/lib/shared/enums.ts#L1-L6

### Assumptions & Initiation Needs
- The existing communication code is well-documented and can be easily understood for refactoring.
- The development team has the necessary knowledge and skills to create and implement the new JavaScript library.
- The VTL pages are already set up and can be accessed for integration.

### PR DESCRIPTION
**Pull request by KevinDavilaDotCMS**
### Proposed Changes
* Moved editor logic to standalone lib

Disclaimer: This PR have comments and dont have testing, descriptions and docs yet. For now we need to discuss about the best approach

For this lib, i move the editor logic (later inside react/next sdk) to editor standalone lib.
For this, we need to extract the logic and create functions that can run in all Javascript frameworks (Angular, Vue, Astro, etc)

So, the types and functions must to use only Browser API

Also, the naming of the lib must to be agnostic to frameworks (Cant use "useXXX" like React)

In the editor-lib we have 2 proposals: object and standalone functions

```
 const dotSdkCustomerActions = {
     someAction: () => {}
}
```

With this approach we gain the advantage to have shorts and clearly functions names because the object name is the glibal description like 

`dotSdkCustomerActions.someAction()`

The other approach is standalone functions like

`export const dotOnSetURL = () => {};`

This approach is simple, because the customer actions doesnt need any comunication between them, but the names are more verbose and must to be clear
I have this proposals
```
export const notifyDotEditorOfSetBounds = () => {};
export const notifyDotEditorOfSetURL = () => {};
```
I really like this because the CustomerActions are a notifier from page to editor.

`export const sendCustomerActionSetURL = () => {};`
The prefix send is common, and then describe exactly that we are doing

`export const dotTriggerSetURL = () => {};`
The prefix trigger is common, but idk if that is exactly what we are doing

```
export const dotOnCustomerActionSetURL = () => {};
or
export const onCustomerActionSerURL = () => {};
```
The prefix on is from BrowserAPI, but the syntax refers a callback, like (when user clicks this, execute this callback), if we dont have problem with this debate, this name is so cool

`export const dotOnSetURL = () => {};`
Same as above, but cut CustomerAction words

### CODE REVIEW

File: core-web/libs/sdk/editor/src/lib/sdk-editor.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                    if (didNodesChanged) {
+                        postMessageToEditor({
+                            action: CUSTOMER_ACTIONS.CONTENT_CHANGE
+                        });
+                    }
+                }
+            }
+        });
+    }
+};
```
Comment (zJaaal) [inline core-web/libs/sdk/editor/src/lib/sdk-editor.ts:83]: Maybe we should use this approach with some difference to make an `init` in order to get the current Host of the dotcms Editor and use it on the post messages as targetOrigin. I think that should be the way to go if we want to catch the target origin and drop the use of `*`. But for me both approaches are right.

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-binary-field/dot-edit-content-binary-field.component.ts [hunk 2]
```diff
@@ -380,9 +383,9 @@ export class DotEditContentBinaryFieldComponent
     private parseCustomMonacoOptions(
         fieldVariables: DotCMSContentTypeFieldVariable[]
     ): Record<string, string> {
-        const { monacoOptions } = getFieldVariablesParsed<{ monacoOptions: string }>(
-            fieldVariables
-        );
+        const { monacoOptions } = getFieldVariablesParsed<{
+            monacoOptions: string;
+        }>(fieldVariables);
```
Comment (fmontes) [inline core-web/libs/edit-content/src/lib/fields/dot-edit-content-binary-field/dot-edit-content-binary-field.component.ts:388]: We're having prettier issues again... not sure what's going on, but there is something going on.

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:1]: Clean up

File: core-web/libs/sdk/editor/README.md
Comment (fmontes) [inline core-web/libs/sdk/editor/README.md:1]: Can't forget about doc, this comment is just so we don't forget :D

File: core-web/libs/sdk/editor/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export * from './lib/actions/customer.actions';
```
Comment (fmontes) [inline core-web/libs/sdk/editor/src/index.ts:1]:
```suggestion
export * from './lib/actions/client.actions';
```

File: core-web/libs/sdk/editor/src/lib/actions/editor.actions.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export enum NOTIFY_CUSTOMER {
```
Comment (fmontes) [inline core-web/libs/sdk/editor/src/lib/actions/editor.actions.ts:1]:
```suggestion
export enum NOTIFY_CLIENT {
```

File: core-web/libs/sdk/editor/src/lib/sdk-editor.ts (hunk 1 above)
Comment (fmontes) [reply core-web/libs/sdk/editor/src/lib/sdk-editor.ts:83]: I don't like any of the two approaches, will be too difficult to explain to the developers how to implement this for their frameworks, I don't have an answer right now, though.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/sdk/editor/src/lib/sdk-editor.ts (hunk 1 above)
Comment (KevinDavilaDotCMS) [reply core-web/libs/sdk/editor/src/lib/sdk-editor.ts:83]: Maybe with other names, something more descriptive

I was thinking that we can use other descriptions for the actions

For example
SET_CONTENTLET = notifies that a contentlet has been hovered and the editor should display the options box.

We could rename this to something like

`showEditOptions()
or
notifyShowEditOptions()`

I think this improves the understanding
And maybe a glossary is necessary to this lib

Comment (fmontes) [reply core-web/libs/sdk/editor/src/lib/sdk-editor.ts:83]: Maybe associate this somehow with the items? like `contentlet.pointerEnter` or something like that.

I think we need to go and check other libraries that are facing this same thing.

Comment (KevinDavilaDotCMS) [reply core-web/libs/sdk/editor/src/lib/sdk-editor.ts:83]: Maybe an object or class separating like
```
export class DotSdkPageEditor{
    editor = {
        updateURL: (url:string) => {},
        didContentChange: () => {},
        eventScrollHandler: () => {},
        isInsideEditor: () => {}
    }

    contentlet = {
        setHoveredContentlet: (e: PointerEvent) => {},
    }
}
```

This is like oher options works, like [Storyblok](https://github.dev/storyblok/storyblok-js-client/blob/3f9fb786a128d4d9ece41ad3c049ec04d2dc6aae/src/index.ts) but the main different is our Editor only manage the Client Actions (for now)

Im looking for separate concerns yet on our editor

Comment (zJaaal) [reply core-web/libs/sdk/editor/src/lib/sdk-editor.ts:83]: One think that we also should think about, is that maybe we will need to make more steps to `add` or `remove` events, so maybe we should provide the `connect/disconnect` events. 

I think we should do more research about this and schedule a meeting, for now it feels like throwing rocks to a window that is very high and we are just hoping we break it in some gracious shot.

File: core-web/libs/edit-content/src/lib/fields/dot-edit-content-binary-field/dot-edit-content-binary-field.component.ts (hunk 2 above)
Comment (oidacra) [reply core-web/libs/edit-content/src/lib/fields/dot-edit-content-binary-field/dot-edit-content-binary-field.component.ts:388]: Verify if you have the prettier file at the root of `core-web`

**Pull request closed without merging**